### PR TITLE
kustomize: Add isolated overlay 

### PR DESCRIPTION
**Pull request by qinqon**
The default kustomize overlay at `config/overlay/default` needs to populate a RoleBinding under
kube-system and a ClusterRole and ClusterRoleBinding this mean that some
resources have to be populated out of the tenant namespace on infra
cluster. 

This change add another kustomize overlay `config/overlay/isolated`
that prevent that by disabling external apiserver authentication
(/metrics and /configz endpoint will not be valid) and disabling zone
and region and tenant nodes. 

Also the `config/overlay/kubevirtci` use as a base `config/overlay/isolated` to run CI with it.

### CODE REVIEW

**Comment by qinqon** [comment]
@mfranczy Looks like this is the issue, do you know if this can open another can of worms

File: cmd/kubevirt-cloud-controller-manager/main.go [hunk 1]
```diff
@@ -43,6 +43,10 @@ func main() {
 		klog.Fatalf("unable to initialize command options: %v", err)
 	}
 
+	// To remove role binding to configmap kube-system:extension-apiserver-authentication-reader
+	// cluster are working fine without it
+	ccmOptions.Authentication.SkipInClusterLookup = true
```
Comment (davidvossel) [inline cmd/kubevirt-cloud-controller-manager/main.go:48]: what behavior in the cloud provider interface actually needs this? Is there a way we can explicitly disable the behavior entirely?

**Comment by mfranczy** [comment]
@davidvossel @qinqon @rhrazdil after reading this discussion https://github.com/kubernetes-sigs/cluster-api-provider-kubevirt/pull/168#discussion_r936015164 I think I get your problem better.

The `ccmOptions.Authentication.SkipInClusterLookup = true` change should not solve your problem.
When you run KubeVirt CCM you should pass a kubeconfig (`--kubeconfig` arg) that points to tenant cluster.
Meaning all `ccmOptions` should influence app behaviour in context of tenat cluster resources, not the infra,
and the lookup for config map should take place on the tenant cluster too.

The CCM itself doesn't have any requirement or need to read config maps from the infrastructure cluster.
If you see this happening then we must have a bug or cloud-provider pkg itself.

From infrastructure cluster we have to have a possibility to create/get/list/delete/watch services, endpoints, get/list vms and update vms labels (it should be a permission to take those actions in a cluster dedicated namespace).

At Kubermatic once we run KubeVirt CCM we pass a kubeconfig to tenant cluster explicitly and kubeconfig to infra cluster is passed over the [`kubeconfig` field of cloud config](https://github.com/kubevirt/cloud-provider-kubevirt/blob/main/pkg/provider/cloud.go#L44) (since KubeVirt in that context is a cloud provider). Next, the infra kubeconfig is used to create a [controller-runtime client](https://github.com/kubevirt/cloud-provider-kubevirt/blob/main/pkg/provider/cloud.go#L132) to act on resources mentioned above.

**Comment by mfranczy** [comment]
Cloud controller manager runs API server so it requires access to `extension-apiserver-authentication` configmap to verify client requests. I assumed that setting `--kubeconfig` that points to  arg is enough. I was wrong. After debugging session with @qinqon we learned that we have to set as well `--kubeconfig-authorization` and `--kubeconfig-authentication`.

If these two parameters are not set then `rest.InClusterConfig` will be used for authorization and authentication. So, if you run KubeVirt CCM on the infra cluster it will try to get the `extension-apiserver-authentication` config map on the infra cluster.

One of the solution to solve this problem is to set those arguments for KubeVirt CCM deployment
```
--kubeconfig=<path-to-tenant-kubeconfig>
--kubeconfig-authorization=<path-to-tenant-kubeconfig>
--kubeconfig-authentication=<path-to-tenant-kubeconfig>
```

File: cmd/kubevirt-cloud-controller-manager/main.go (hunk 1 above)
Comment (mfranczy) [reply cmd/kubevirt-cloud-controller-manager/main.go:48]: We don't need that option. We found why you had a problem with CCM trying to reach the config map on the infra cluster.
As described here https://github.com/kubevirt/cloud-provider-kubevirt/pull/141#issuecomment-1231532936.

**Comment by qinqon** [comment]
> Cloud controller manager runs API server so it requires access to `extension-apiserver-authentication` configmap to verify client requests. I assumed that setting `--kubeconfig` that points to tenant cluster is enough. I was wrong. After debugging session with @qinqon we learned that we have to set as well `--authorization-kubeconfig` and `--authentication-kubeconfig`.
> 
> If these two parameters are not set, `rest.InClusterConfig` will be used to build the client for the API server. So, if you run KubeVirt CCM on infra cluster without these parameters pointing to the tenant cluster, it will try to get the extension authentication config map on the infra cluster.
> 
> One of the solution to solve this problem is to set those arguments for KubeVirt CCM deployment
> 
> ```
> --kubeconfig=<path-to-tenant-kubeconfig>
> --authorization-kubeconfig=<path-to-tenant-kubeconfig>
> --authentication-kubeconfig=<path-to-tenant-kubeconfig>
> ```

Have rework the PR with what we found so we don't have to "disable" features.

**Comment by mfranczy** [comment]
Before we merge that PR I would like to go through controller manager API server implementation carefully to double check that it's a right approach. I should have answers soon.

**Comment by mfranczy** [comment]
@qinqon I have my answers.

The CCM API expose three endpoints
* /healthz (doesn't need authentication)
* /metrics
* /configz (dumps whole configuration of CCM)

By passing `--authentication-skip-lookup` option (the `Authentication.SkipInClusterLookup`) you eliminate access to `extension-apiserver-authentication` config map on the KubeVirt cluster. However, that might disable validation of http(s) requests meaning all users will be anonymous for the API server in the CCM.

If you pass `--authentication-kubeconfig` to the API server of tenant cluster then requests would be validated against the certificate that exists in `extension-apiserver-autentication` config map on tenant cluster. This is wrong in case you have some applications running on KubeVirt infra cluster that try to reach `/configz` or `/metrics` endpoints. Although I understand that's also not your case.

Because community can have different ways of deploying the CCM, I would suggest that we don't remove RBAC to `extension-apiserver-authorization` from default config. IMO the best would be to add another config that cover your case. What do you think?

**Comment by qinqon** [comment]
> @qinqon I have my answers.
> 
> The CCM API expose three endpoints
> * /healthz (doesn't need authentication)
> * /metrics
> * /configz (dumps whole configuration of CCM)
> 
> By passing `--authentication-skip-lookup` option (the `Authentication.SkipInClusterLookup`) you eliminate access to `extension-apiserver-authentication` config map on the KubeVirt cluster. However, that might disable validation of http(s) requests meaning all users will be anonymous for the API server in the CCM.
> 
> If you pass `--authentication-kubeconfig` to the API server of tenant cluster then requests would be validated against the certificate that exists in `extension-apiserver-autentication` config map on tenant cluster. This is wrong in case you have some applications running on KubeVirt infra cluster that try to reach `/configz` or `/metrics` endpoints. Although I understand that's also not your case.
> 
> Because community can have different ways of deploying the CCM, I would suggest that we don't remove RBAC to `extension-apiserver-authorization` from default config. IMO the best would be to add another config that cover your case. What do you think?

Thanks for the investigation! So we can still has /healthz with --authentication-skip-lookup=true ? Is this this is the case we can just delegate on how community start the manager to decide how to make it work, at the end the ARGS ar passed to apisever, I think.

**Comment by mfranczy** [comment]
> So we can still has /healthz with --authentication-skip-lookup=true

yes, because healthz is on the allowed list as described:

> // This allows the kubelet to always get health and readiness without causing an authorization check.
		// This field can be cleared by callers if they don't want this behavior.
		AlwaysAllowPaths: []string{"/healthz", "/readyz", "/livez"},

> Is this this is the case we can just delegate on how community start the manager to decide how to make it work, 

After reading the code I think community does that well. It's just unusual scenario where extension API runs on a cluster but doesn't want to read root certificate to validate requests because of the isolation. Maybe you can pass somehow the certificate to the tenant namespace (I didn't check that). Although that would mean copying the certifcate on each tenant cluster creation.

> at the end the ARGS ar passed to apisever

yes, the `CompletedConfig` is passed to the API server options of CCM.

**Comment by qinqon** [comment]
> > So we can still has /healthz with --authentication-skip-lookup=true
> 
> yes, because healthz is on the allowed list as described (you can extend the list):
> 
> > // This allows the kubelet to always get health and readiness without causing an authorization check.
> > // This field can be cleared by callers if they don't want this behavior.
> > AlwaysAllowPaths: []string{"/healthz", "/readyz", "/livez"},
> 
> > Is this this is the case we can just delegate on how community start the manager to decide how to make it work,
> 
> After reading the code I think community does that well. It's just unusual scenario where extension API runs on a cluster but it's not allowed to read client-ca-file certificate to validate requests because of the isolation. Maybe you can pass somehow the certificate to the tenant namespace (I didn't check that). Although that would mean copying/generating the certifcate on each tenant cluster creation. Another solution would be to use different authentication method than client certificate authentication.
> 
> > at the end the ARGS ar passed to apisever
> 
> yes, the `CompletedConfig` is passed to the API server options of CCM.

I am going to change this PR to add another kustomize overlay that will remove the kube-system role binding and add the --authentication-skip-lookup=true flag to deployment so the community can decide how to deploy, the golang code will not be changed.

**Review by mfranczy — APPROVED** [review]
/lgtm
/approve

**Pull request merged**